ouroboros-consensus-protocol/src/Ouroboros/Consensus/Protocol/Praos.hs

```diff
@@ -0,0 +1,684 @@
+            toCBOR praosStateEpochNonce,
+            toCBOR praosStateLabNonce,
+            toCBOR praosStateLastEpochBlockNonce
+          ]
+
+  decode =
+    decodeVersion
+      [(serialisationFormatVersion1, Decode decodePraosState1)]
+    where
+      decodePraosState1 = do
+        enforceSize "PraosState" 7
+        PraosState
+          <$> fromCBOR
+          <*> fromCBOR
+          <*> fromCBOR
+          <*> fromCBOR
+          <*> fromCBOR
+          <*> fromCBOR
+          <*> fromCBOR
+
+-- | Ticked 'PraosState'
+data instance Ticked (PraosState c) = TickedPraosState
+  { tickedPraosStateChainDepState :: PraosState c,
+    tickedPraosStateLedgerView :: Ticked (Views.LedgerView c)
+  }
+
+-- | Errors which we might encounter
+data PraosValidationErr c
+  = VRFKeyUnknown
+      !(KeyHash SL.StakePool c) -- unknown VRF keyhash (not registered)
+  | VRFKeyWrongVRFKey
+      !(KeyHash SL.StakePool c) -- KeyHash of block issuer
+      !(SL.Hash c (SL.VerKeyVRF c)) -- VRF KeyHash registered with stake pool
+      !(SL.Hash c (SL.VerKeyVRF c)) -- VRF KeyHash from Header
+  | VRFKeyBadNonce
+      !SlotNo -- Slot used for VRF calculation
+      !Nonce -- Epoch nonce used for VRF calculation
+      !(VRF.CertifiedVRF (VRF c) UnifiedVRF) -- VRF calculated nonce value
+  | VRFLeaderValueTooBig (VRF.OutputVRF (VRF c)) Rational ActiveSlotCoeff
+  | KESBeforeStartOCERT
+      !KESPeriod -- OCert Start KES Period
+      !KESPeriod -- Current KES Period
+  | KESAfterEndOCERT
+      !KESPeriod -- Current KES Period
+      !KESPeriod -- OCert Start KES Period
+      !Word64 -- Max KES Key Evolutions
+  | CounterTooSmallOCERT
+      !Word64 -- last KES counter used
+      !Word64 -- current KES counter
+  | -- | The KES counter has been incremented by more than 1
+    CounterOverIncrementedOCERT
+      !Word64 -- last KES counter used
+      !Word64 -- current KES counter
+  | InvalidSignatureOCERT
+      !Word64 -- OCert counter
+      !KESPeriod -- OCert KES period
+      !String -- DSIGN error message
+  | InvalidKesSignatureOCERT
+      !Word -- current KES Period
+      !Word -- KES start period
+      !Word -- expected KES evolutions
+      !String -- error message given by Consensus Layer
+  | NoCounterForKeyHashOCERT
+      !(KeyHash 'BlockIssuer c) -- stake pool key hash
+  deriving (Generic)
+
+deriving instance PraosCrypto c => Eq (PraosValidationErr c)
+
+deriving instance PraosCrypto c => NoThunks (PraosValidationErr c)
+
+deriving instance PraosCrypto c => Show (PraosValidationErr c)
+
+instance PraosCrypto c => ConsensusProtocol (Praos c) where
+  type ChainDepState _ = PraosState c
+  type IsLeader _ = PraosIsLeader c
+  type CanBeLeader _ = PraosCanBeLeader c
+  type SelectView _ = PraosChainSelectView c
+  type LedgerView _ = Views.LedgerView c
+  type ValidationErr _ = PraosValidationErr c
+  type ValidateView _ = PraosValidateView c
+
+  protocolSecurityParam = praosSecurityParam . praosParams
+
+  checkIsLeader
```

Comment:
I have yet to review the rest of the file from here down. In particular we'll want to check the new unified VRF against the researchers' spec.